reactive/kubernetes_master.py

```diff
@@ -2818,3 +2819,32 @@ def api_server_stopped():
     aws_iam = endpoint_from_flag('endpoint.aws-iam.available')
     if aws_iam:
         aws_iam.set_api_server_status(False)
+
+
+@when('cluster-dns.joined')
+def send_clusterdns_domain():
+    """Send domain to the Cluster DNS interface."""
+    cluster_dns = endpoint_from_flag('cluster-dns.joined')
+    dns_domain = hookenv.config('dns_domain')
+
+    hookenv.log('Sending dns_domain: {}'.format(dns_domain))
+    cluster_dns.send_domain(dns_domain)
+
+    set_flag('kubernetes-master.need-cluster-dns')
+    clear_flag('cluster-dns.joined')
+
+
+@when('cluster-dns.ready')
+@when('kubernetes-master.need-cluster-dns')
+@when('kube-control.connected')
+def set_charmed_clusterdns():
+    """Send new ClusterDNS information to the worker"""
+    kube_control = endpoint_from_flag('kube-control.connected')
+    cluster_dns = endpoint_from_flag('cluster-dns.ready')
+    dns_ip = cluster_dns.get_ip()
+    dns_domain = hookenv.config('dns_domain')
+
+    hookenv.log('Sending dns_ip: {}'.format(dns_ip))
+    kube_control.set_dns(53, dns_domain, dns_ip, True)
+
+    clear_flag('kubernetes-master.need-cluster-dns')
```

Comment:
What if `dns_domain` changes? What if `dns_ip` changes?

Reply:
How could an IP of a service change without restarting the whole service?

Reply:
No idea. It's generally not supposed to happen. But I've seen the occasional bug report where the kube-dns service IP changed unexpectedly and the user had to recreate all of their pods to pick up the new DNS IP. It's rare, and I have no idea how it happens, but it happens. Maybe it will never happen with the coredns k8s charm, but we should be prepared for that possibility and handle it appropriately.

Reply:
I think juju would also get lost if the service underneath the k8s charm changes IP. But I will ask the juju k8s team.